libr/crypto/p/crypto_rc6.c

```diff
@@ -0,0 +1,222 @@
+
+	int i, off;
+
+	for (i = 0, off = 0; i < c; i++) {
+		L[i] = ((key[off++] & 0xff)) |
+				((key[off++] & 0xff) << 8) |
+				((key[off++] & 0xff) << 16) |
+				((key[off++] & 0xff) << 24);
+	}
+
+	(state->S)[0] = Pw;
+	for (i = 1; i < t; i++) {
+		(state->S)[i] = (state->S)[i-1] + Qw;
+	}
+
+	for (i = 0; i < v; i++) {
+		A = (state->S)[k] = ROTL(((state->S)[k] + A + B), 3);
+		B = L[j] = ROTL(L[j] + A + B, A + B);
+		k = (k + 1) % t;
+		j = (j + 1) % c;
+	}
+	
+	state->key_size = keylen/8;
+	return true;
+}
+
+static void rc6_encrypt(struct rc6_state *const state, const ut8 *inbuf, ut8 *outbuf) {
+	ut32 t, u;
+	ut32 aux;
+	ut32 data[BLOCK_SIZE/4];
+	int i;
+	int off = 0;
+	for (i = 0; i < BLOCK_SIZE / 4; i++) {
+		data[i] = ((inbuf[off++] & 0xff)) |
+					((inbuf[off++] & 0xff) << 8) |
+					((inbuf[off++] & 0xff) << 16) |
+					((inbuf[off++] & 0xff) << 24);
+	}
+
+	ut32 A = data[0], B = data[1], C = data[2], D = data[3];
+
+	// S is key
+	B = B + (state->S)[0];
+	D = D + (state->S)[1];
+
+	for (i = 1; i <= r; i++) {
+		t = ROTL(B * (2 * B + 1), 5);		//lgw == 5
+		u = ROTL(D * (2 * D + 1), 5);
+		A = ROTL(A ^ t, u) + (state->S)[2 * i];
+		C = ROTL(C ^ u, t) + (state->S)[2 * i + 1];
+
+		aux = A;
+		A = B;
+		B = C;
+		C = D;
+		D = aux;
+	}
+
+	A = A + (state->S)[2*(r+1)];
+	C = C + (state->S)[2*(r+1)+1];
+	data[0] = A;
+	data[1] = B;
+	data[2] = C;
+	data[3] = D;
+
+	for (i = 0; i < BLOCK_SIZE; i++) {
+		outbuf[i] = (ut8)((data[i / 4] >> (i % 4) * 8) & 0xff);
+	}
+}
+
+static void rc6_decrypt(struct rc6_state *const state, const ut8 *inbuf, ut8 *outbuf) {
+	ut32 t, u;
+	ut32 aux;
+	ut32 data[BLOCK_SIZE / 4];
+	int i;
+	int off = 0;
+
+	for (i = 0; i < BLOCK_SIZE / 4; i++) {
+		data[i] = ((inbuf[off++] & 0xff)) |
+					((inbuf[off++] & 0xff) << 8) |
+					((inbuf[off++] & 0xff) << 16) |
+					((inbuf[off++] & 0xff) << 24);
+	}
+
+	ut32 A = data[0], B = data[1], C = data[2], D = data[3];
+
+	C = C - (state->S)[2 * (r + 1) + 1];
+	A = A - (state->S)[2 * (r + 1)];
+
+	for (i = r; i >= 1; i--) {
+		aux = D;
+		D = C;
+		C = B;
+		B = A;
+		A = aux;
+
+		u = ROTL(D * (2 * D + 1), 5);
+		t = ROTL(B * (2 * B + 1), 5);
+		C = ROTR(C - (state->S)[2 * i + 1], t) ^ u;
+		A = ROTR(A - (state->S)[2 * i], u) ^ t;
+	}
+
+	D = D - (state->S)[1];
+	B = B - (state->S)[0];
+
+	data[0] = A;
+	data[1] = B;
+	data[2] = C;
+	data[3] = D;
+
+	for (i = 0; i < BLOCK_SIZE; i++) {
+		outbuf[i] = (ut8)((data[i / 4] >> (i % 4) * 8) & 0xff);
+	}
+}
+
+static struct rc6_state st;
+
+static int rc6_set_key(RCrypto *cry, const ut8 *key, int keylen, int mode, int direction) {
+	return rc6_init (&st, key, keylen, direction);
+}
+
+static int rc6_get_key_size(RCrypto *cry) {
+	return st.key_size;
+}
+
+static bool rc6_use(const char *algo) {
+	return !strcmp (algo, "rc6");
+}
+
+static int update(RCrypto *cry, const ut8 *buf, int len) {
+	if (len % BLOCK_SIZE != 0) {
+		return false;
+	}
+
+	const int diff = (BLOCK_SIZE - (len % BLOCK_SIZE)) % BLOCK_SIZE;
```

Comment:
actually yes diff as of now would be always 0. but what I intended to do was in case the input is not multiple of BLOCK_SIZE, it should be padded. here it is padding null bytes. I will fix this.